server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -145,6 +146,20 @@ protected BulkShardResponse newResponseInstance(StreamInput in) throws IOExcepti
         return new BulkShardResponse(in);
     }
 
+    @Override
+    protected void shardOperationOnPrimary(
```

Comment:
I guess that you decided to do the listener call here instead of `dispatchedShardOperationOnPrimary` to avoid dispatching in the thread pool for short lived tasks?

Reply:
Indeed! Forgot to mention I briefly discussed this yesterday with @henningandersen as well, and noticed if we had put it in `dispatchedShardOperationOnPrimary` we would unnecessarily switch from transport thread -> write thread pool -> unhollow (currently generic) -> write thread pool.

With the current PR, we simplify and switch as follows: transport thread -> unhollow -> write thread pool. So the actual bulk is only queued/executed once in the write thread pool.